### PR TITLE
Add a new question for Consul discovery & configuration

### PR DESCRIPTION
**Pull request by PierreBesson**
This is quite a big PR, I hope you can help me test it. It worked nicely in all cases I tried. Also we need to take good care that this is not breaking eureka in any way or causing problem when doing the upgrade.

It contains:
- Changes to the server subgen to add a new question for service discovery
- Select between spring-cloud-starter-eureka and spring-cloud-consul build dependencies
- Add the necessary properties to connect to Consul.
- Add a consul.yml to start a dev consul server. You will notice that it uses `pbesson/consul-config-loader` docker image, I intend to transfer ownership of this to the jhipster org before merging. This is a really simple container that helps dealing with consul configuration. Otherwise people would have to manually type curl commands to load their YAML files to consul. For now it is a bash script but I would like to rewrite it as a nodejs script so that I can also publish it on npm for people that don't want to use docker.
- I also updated the docker-compose subgen. It was important for me to have this in the same PR to see if it everything was working correctly.

I also made Hazelcast clustering work when doing a `docker-compose scale`, the solution was enabling the `prefer-ip-address` property.

### CODE REVIEW

**Comment by xetys** [comment]
first look: good work! I will play with these settings as I get time for that...

as in the mailing list I am thinkig supporting option "no" to discovery is a good idea...I think of ribbon, hystrix and feign, which may just don't anymore...I fear having this option inside without any discovery will lead to failures....but i'll try that first

**Comment by PierreBesson** [comment]
@xetys, Actually it is theoretically possible to use ribbon even without a service discovery backend.
`To configure Ribbon with a fixed list of physical servers you can simply set <client>.ribbon.listOfServers to a comma-separated list of physical addresses (or hostnames), where <client> is the ID of the client.`

I have not tested this but I think it means that you can use Feign, Hystrix, etc,... by just defining serviceName.ribbon.listOfServers=server1.com, server2.com, this might be useful for people that use their own solution for service discovery.

File: generators/server/templates/src/main/docker/config/git2consul.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "version": "1.0",
+  "repos" : [{
+    "name" : "config",
+    "url" : "https://github.com/PierreBesson/test-registry.git",
```
Comment (deepu105) [inline generators/server/templates/src/main/docker/config/git2consul.json]: Is this correct?

File: generators/server/prompts.js [hunk 2]
```diff
@@ -101,6 +101,31 @@ function askForServerSideOpts() {
         },
         {
             when: function (response) {
+                return applicationType === 'gateway' || applicationType === 'microservice' || applicationType === 'uaa';
+            },
+            type: 'list',
+            name: 'serviceDiscoveryType',
+            message: function (response) {
+                return getNumberedQuestion('Which Service Discovery and Configuration solution would you like to use?', applicationType === 'gateway' || applicationType === 'microservice');
```
Comment (deepu105) [inline generators/server/prompts.js]: You missed the uaa option in the condition here

**Comment by xetys** [comment]
I know, that is theoretically possible, but will it simply work for our microservices setup if we just remove discovery?

**Comment by pascalgrimaud** [comment]
@PierreBesson : FYI, there are some errors on Travis build

**Comment by agoncal** [comment]
I played a bit with this PR. 

I created 3 microservices, one gateway, killed/start/stopped those services.... and Consul was doing its job as well as the Gateway displaying the living microservices.

The only problem I found is, when Consul is down, and you start a service, the exception is not clear (see stacktrace below). I would have liked to see something clearer such "Cannot locate Consul. Make sure you have Consul started. Shutting down gracefully" instead of just "Fail fast is set and there was an error reading configuration from consul":

`2016-09-21 15:22:51.968 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:52.977 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:54.083 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:55.297 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:56.633 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:58.103 ERROR 9431 --- [  restartedMain] o.s.c.c.c.ConsulPropertySourceLocator    : Fail fast is set and there was an error reading configuration from consul.
2016-09-21 15:22:58.110  WARN 9431 --- [  restartedMain] o.s.boot.SpringApplication               : Error handling failed (ApplicationEventMulticaster not initialized - call 'refresh' before multicasting events via the context: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@6c27834a: startup date [Thu Jan 01 01:00:00 CET 1970]; parent: org.springframework.context.annotation.AnnotationConfigApplicationContext@7566b00)
2016-09-21 15:22:58.127 ERROR 9431 --- [  restartedMain] o.s.boot.SpringApplication               : Application startup failed

com.ecwid.consul.transport.TransportException: java.net.ConnectException: Connection refused
	at com.ecwid.consul.transport.AbstractHttpTransport.executeRequest(AbstractHttpTransport.java:91)
	at com.ecwid.consul.transport.AbstractHttpTransport.makeGetRequest(AbstractHttpTransport.java:50)
	at com.ecwid.consul.v1.ConsulRawClient.makeGetRequest(ConsulRawClient.java:81)
	at com.ecwid.consul.v1.kv.KeyValueConsulClient.getKVValues(KeyValueConsulClient.java:150)
	at com.ecwid.consul.v1.ConsulClient.getKVValues(ConsulClient.java:404)
	at org.springframework.cloud.consul.config.ConsulPropertySource.init(ConsulPropertySource.java:64)
	at org.springframework.cloud.consul.config.ConsulPropertySourceLocator.create(ConsulPropertySourceLocator.java:135)
	at org.springframework.cloud.consul.config.ConsulPropertySourceLocator.locate(ConsulPropertySourceLocator.java:113)
	at org.springframework.cloud.consul.config.ConsulPropertySourceLocator$$FastClassBySpringCGLIB$$b35ebf8.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:720)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:74)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:276)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:157)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:101)
	at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:118)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:655)
	at org.springframework.cloud.consul.config.ConsulPropertySourceLocator$$EnhancerBySpringCGLIB$$319f733c.locate(<generated>)
	at org.springframework.cloud.bootstrap.config.PropertySourceBootstrapConfiguration.initialize(PropertySourceBootstrapConfiguration.java:89)
	at org.springframework.boot.SpringApplication.applyInitializers(SpringApplication.java:633)
	at org.springframework.boot.SpringApplication.prepareContext(SpringApplication.java:347)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:311)
	at com.mycompany.myapp.MoneApp.main(MoneApp.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:120)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:179)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:328)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:612)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:447)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:884)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:220)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:164)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:139)
	at com.ecwid.consul.transport.AbstractHttpTransport.executeRequest(AbstractHttpTransport.java:75)
	... 29 common frames omitted

`

File: generators/server/templates/src/main/docker/config/git2consul.json (hunk 1 above)
Comment (PierreBesson) [inline generators/server/templates/src/main/docker/config/git2consul.json]: This is more like a sample config, in before merging, I will change this to `url: jhipster/generator-jhipster.git` and `source_root": "/generators/.../consul-config/`.

**Comment by pascalgrimaud** [comment]
I'm fixing travis build and will PR to your branch @PierreBesson 
Then, you'll have to rebase, because there is conflicts

File: generators/docker-compose/prompts.js [hunk 3]
```diff
@@ -167,8 +168,69 @@ function askForElk() {
...
+    else if(serviceDiscoveryEnabledApps.every(app => app.serviceDiscoveryType === 'eureka')){
+        this.serviceDiscoveryType = 'eureka';
+        this.log(chalk.green('JHipster registry detected as the service discovery and configuration provider used by your apps'));
+        done();
+    }
+    else {
+        this.log(chalk.yellow('Unable to determine the service discovery and configuration provider to use from your apps configuration.'));
+        this.log('Your service discovery enabled apps:');
+        serviceDiscoveryEnabledApps.forEach(function(app){
+            this.log('  - ' + app.baseName + ' ('  + app.serviceDiscoveryType + ')');
```
Comment (deepu105) [inline generators/docker-compose/prompts.js]: you could do `` `  -${app.baseName} (${app.serviceDiscoveryType})` `` since we are on es6

**Comment by pascalgrimaud** [comment]
Good job @PierreBesson ! Just tested and it works fine.
I have one question: consul need to be started, otherwise the `mvn test` failed. Is it normal ?

File: generators/server/templates/src/main/resources/config/_application.yml [hunk 4]
```diff
@@ -11,23 +11,23 @@
...
-
 <%_ } _%>
 <%_ if (applicationType == 'gateway') { _%>
 zuul:
     host:
         connect-timeout-millis: 5000
         socket-timeout-millis: 10000
-
+    <%_ if (serviceDiscoveryType == 'consul') { _%>
+    ignoredServices: 'consul,gateway' # Prevent Consul and the "<%= baseName %>" app itself from having a Zuul route automatically created for them
```
Comment (pascalgrimaud) [inline generators/server/templates/src/main/resources/config/_application.yml]: shouldn't it be <%= baseName %> instead of 'gateway' ?

**Comment by ruddell** [comment]
@PierreBesson Nice work!

@pascalgrimaud Adding a bootstrap.yml with  `spring.cloud.consul.config.enabled: false` fixes the test issue.  Similar to #4203 but with `consul` in the property

**Comment by pascalgrimaud** [comment]
@ruddell : yes, you're right! this `bootstrap.yml` with this config, should be add too

File: generators/server/templates/src/main/resources/config/_application.yml (hunk 4 above)
Comment (PierreBesson) [inline generators/server/templates/src/main/resources/config/_application.yml]: You are totally right. I'm doing the fix right now.

**Comment by PierreBesson** [comment]
Thanks a lot @ruddell ! I just fixed the issue. Maybe my PR will finally be ready to merge ?

**Comment by pascalgrimaud** [comment]
LGTM

**Comment by xetys** [comment]
If this is OK, I would test this branch on my JHipster uaa setup to be sure. This will be tonight or tomorrow before noon

**Comment by PierreBesson** [comment]
@xetys yes please test with the UAA, then we will be able to merge this. Thank you very much.

File: generators/server/prompts.js (hunk 2 above)
Comment (deepu105) [inline generators/server/prompts.js]: @PierreBesson the `uaa` check is still missing here

File: generators/server/prompts.js [hunk 5]
```diff
@@ -101,6 +101,31 @@ function askForServerSideOpts() {
...
+                {
+                    value: 'eureka',
+                    name: 'JHipster Registry (using Eureka and Spring Cloud Config)'
+                },
+                {
+                    value: 'consul',
+                    name: 'Consul (using Spring Cloud Consul)'
+                },
+                {
+                    value: 'no',
```
Comment (deepu105) [inline generators/server/prompts.js]: can we use false instead of no please so that its aligned with other options we have

**Comment by deepu105** [comment]
@PierreBesson sorry to add more comments, hope you dont mind

File: generators/docker-compose/templates/_docker-compose.yml [hunk 6]
```diff
@@ -14,11 +14,24 @@ services:
             service: kafka
 <%_ } _%>
 <%_ if (gatewayNb + microserviceNb > 0) { _%>
+    <%_ if (serviceDiscoveryType == 'eureka') { _%>
```
Comment (xetys) [inline generators/docker-compose/templates/_docker-compose.yml]: just in mind, I was told to use === over == everytime i can do...should we then enforce this policy for all of us :smile: ?

Comment (deepu105) [inline generators/docker-compose/templates/_docker-compose.yml]: yes `===` always.

File: generators/server/prompts.js (hunk 5 above)
Comment (PierreBesson) [inline generators/server/prompts.js]: done

File: generators/docker-compose/templates/_docker-compose.yml (hunk 6 above)
Comment (PierreBesson) [inline generators/docker-compose/templates/_docker-compose.yml]: @xetys You are right and I must say that I have been influenced by all those bad `==` in EJS templates (I used to think that triple equals were not supported in templates), we should get rid of all of those once and for all but maybe not today.

**Comment by pascalgrimaud** [comment]
Is it ok with your tests @xetys ?
@PierreBesson: you need to rebase again...

File: generators/server/prompts.js [hunk 7]
```diff
@@ -101,6 +101,31 @@ function askForServerSideOpts() {
         },
         {
             when: function (response) {
+                return applicationType === 'gateway' || applicationType === 'microservice' || applicationType === 'uaa';
```
Comment (xetys) [inline generators/server/prompts.js]: if you choose uaa for microservices, the prompt is first asking "uaa?", then "discovery" and then some uaa related question to find the uaa for uaaBasename....this is confusing, please move the registry question after the uaa, so it asks me for uaa directory before service discovery

**Comment by xetys** [comment]
Ok, here is my short review, 

I found 2 things must to be fixed...if I spammed a bit, sorry, I am getting familiar with GitHubs new reviewing functions :laughing: 

I also tested feign clients with uaa...works like charm...no...it works even better then it ever worked with eureka :smile: 

great work! I hope this gets merged and released as soon as possible...


one closing thought about the consul agent....it runs in dev mode, which is **not** recommended for production use by HashiCorp...do we keep it like that?

**Comment by xetys** [comment]
one more thing is I noticed....I tested the behavior of registering / unregistering services by stoping and starting docker containers...and this is producing doubletes in registry. While the IP is keeps correct for all instances (no failing requests as for eureka), but the list is growing over time since each service start with a new random uuid in "instance-id"....

not sure if this is good

**Comment by PierreBesson** [comment]
@xetys, thanks for the review I will fix the issues you pointed out.
About the consul agent, the config I provided is only for dev. For prod, you will need to have a consul cluster.
I also noticed that stopping/killing the apps docker containers was not unregistering the apps in consul. So the issue should be investigated, but it is not a blocker as it is docker specific.

**Comment by PierreBesson** [comment]
@xetys I have reordered the prompts for the uaa questions. However I did not fix the uaa healthcheck. Can you please do it as a PR later.
@pascalgrimaud Sorry I removed a kubernetes test (https://github.com/jhipster/generator-jhipster/pull/4165/commits/ba5449d58e804dcde8834e7b420636afbffefcaf) as it was failing with my PR and I could not trace back the reason it was failing.

I think the PR is now ready to merge.

**Comment by pascalgrimaud** [comment]
Very good job @PierreBesson !!
No problem for the test, I ll have a look

**Comment by xetys** [comment]
I will have 7 hours of coach with full charged notebook...I will fix that
today

сб, 1 Окт 2016 г., 9:57 Pascal Grimaud <notifications@github.com>:

**Pull request merged**